packages/flutter_tools/lib/src/runner/target_devices.dart

```diff
@@ -492,9 +492,18 @@ class TargetDevicesWithExtendedWirelessDeviceDiscovery extends TargetDevices {
       );
       if (devices.length == 1) {
         Device? matchedDevice = devices.first;
+
+        if (matchedDevice is IOSDevice) {
+          final bool devModeEnabled = await matchedDevice.isDeveloperModeEnabled();
+          if (!devModeEnabled) {
+            throwToolExit('');
```

Comment:
This tool exit is another case of an unexpected side-effect of checking for `findAllTargetDevices`.  What if the caller doesn't care if the dev mode is enabled and instead just wants a list of good, ready devices? Or it wants to fall back to macOS/web if there are no ready iOS devices?

Instead you should `toolExit` in the spot where the command knows enough to understand this is a problem and can't continue.